### PR TITLE
Feature/support for configuring tracked entity type attribute

### PR DESCRIPTION
**Pull request by john-melin**
- Refactored ProgramAttributeRow from tracker-propgrams into a generic attribute-row
- Added move order functionality to trackedEntityAttributes editor
- Added support for trackedEntityTypeAttribute in trackedEntityType
- Linting and Refactoring of related files

### CODE REVIEW

File: src/EditModel/event-program/tracker-program/assign-tracked-entity-attributes/AssignAttributes.js [hunk 1]
```diff
@@ -1,23 +1,25 @@
-import React, { PropTypes } from 'react';
+import React from 'react';
+import PropTypes from 'prop-types';
+
 import { bindActionCreators } from 'redux';
 import { connect } from 'react-redux';
 import { get, getOr, __ } from 'lodash/fp';
```
Comment (varl) [inline src/EditModel/event-program/tracker-program/assign-tracked-entity-attributes/AssignAttributes.js]: What's the `__` for btw?

File: src/EditModel/event-program/tracker-program/assign-tracked-entity-attributes/AssignAttributes.js [hunk 2]
```diff
@@ -152,17 +98,26 @@ function AssignAttributes(props, { d2 }) {
 
     // Assign existing attributes
     assignedItemStore.setState(
-        props.items.map(a => a.trackedEntityAttribute.id),
+        props.assignedAttributes.map(a => a.trackedEntityAttribute.id),
     );
 
+    const onMoveAttributes = (newAttributesOrderIds) => {
+        assignedItemStore.setState(newAttributesOrderIds);
+        // need to update this.props.assignedAttributes to reflect new order in epics
```
Comment (varl) [inline src/EditModel/event-program/tracker-program/assign-tracked-entity-attributes/AssignAttributes.js]: Trivial, but consider adding `TODO` to such comments. IDEs like IntelliJ automatically keep track of TODOs in comments and for the rest they are easily searchable with tools like `ack` and `grep`.

File: src/EditModel/event-program/tracker-program/assign-tracked-entity-attributes/AssignAttributes.js [hunk 3]
```diff
@@ -152,17 +98,26 @@ function AssignAttributes(props, { d2 }) {
...
+            <ProgramAttributeRow
                 key={programAttribute.id}
-                programAttribute={programAttribute}
-                onEditProgramAttribute={props.onEditProgramAttribute}
+                displayName={programAttribute.trackedEntityAttribute.displayName}
+                attribute={programAttribute}
+                onEditAttribute={props.onEditProgramAttribute}
+                isDateValue={programAttribute.trackedEntityAttribute.valueType === 'DATE'}
+                isUnique={programAttribute.trackedEntityAttribute.unique}
+                hasOptionSet={!!programAttribute.trackedEntityAttribute.optionSet}
```
Comment (varl) [inline src/EditModel/event-program/tracker-program/assign-tracked-entity-attributes/AssignAttributes.js]: is the `ProgramAttributeRow` used elsewhere? Looks like it's reading a lot of props from the same object, which is also passed in, so looks the values could be spread on the inside of `ProgramAttributeRow` component?

File: src/config/field-overrides/tracked-entity-type/AssignTrackedEntityTypeAttributes.component.js [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        assignedAttributesStore: Store.create(),
+        filterText: '',
+        assignedAttributes: this.props.model.trackedEntityTypeAttributes || [],
+    };
+
+    componentDidMount() {
+        this.context.d2.models.trackedEntityAttribute.list({
+            level: 1,
+            paging: false,
+            fields: ['id,displayName,valueType,unique,optionSet,mandatory,searchable,displayInList'].join(','),
```
Comment (varl) [inline src/config/field-overrides/tracked-entity-type/AssignTrackedEntityTypeAttributes.component.js]: Uh?

This evaluates to the same string which is in the array.

```
['id,displayName,valueType,unique,optionSet,mandatory,searchable,displayInList'].join(',') === 'id,displayName,valueType,unique,optionSet,mandatory,searchable,displayInList'
// true
```

File: src/EditModel/event-program/tracker-program/assign-tracked-entity-attributes/AssignAttributes.js (hunk 1 above)
Comment (john-melin) [reply src/EditModel/event-program/tracker-program/assign-tracked-entity-attributes/AssignAttributes.js]: No idea, something birk used.

File: src/EditModel/event-program/tracker-program/assign-tracked-entity-attributes/AssignAttributes.js (hunk 3 above)
Comment (john-melin) [reply src/EditModel/event-program/tracker-program/assign-tracked-entity-attributes/AssignAttributes.js]: Ah, yes. It also used in trackedEntityType. Spreading would make it non-generic.

Comment (varl) [reply src/EditModel/event-program/tracker-program/assign-tracked-entity-attributes/AssignAttributes.js]: Ok.

**Comment by varl** [comment]
LGTM 👍

**Pull request merged**